### ISSUE TITLE
Benchmarks of WASM and WebWorkers

### ISSUE THREAD

**Issue by YoranBrondsema** [nodeca/pica#88]
Hi,

First, thanks for providing this library. We've been using the 2.x version but I'm excited to upgrade to 3.x. Great work on that!

As far as I can see, the benchmarks in https://github.com/nodeca/pica/tree/master/benchmark only test the JS version. However, it would be interesting to extend them to test each scenario with the different feature options (WebWorkers, WebAssembly, createImageBitmap, pure JS).

I think these numbers, while highly dependent on the browser of course, would be interesting for development purposes naturally but also for a user to make a good choice of these options for their use cases. For instance, I'm not 100% sure about this but if a user's use case is only to resize small images, the overhead of creating WebWorkers might be higher than the gain in performance. If the benchmarks show this, the user will be able to make a more educated decision.

I don't mind working on this, just wanted to know your thoughts before I get started!

**Comment by puzrin** [comment]
I have no plans to create official benchmarker - don't like programming for browsers. I think, existing web demo is enougth to do estimates for user cases, it shows total resize time.

`benchmark/` is for internal needs, and wasm test not exists because node.js v8 required. It will be released this month, and i'll add more measurements then. On average, wasm is only 10-20% faster, because we did JS code very well :). Next speed gap will appear only when SIMD come to wasm. Now i use pica to investigate how to write compact wasm modules, just for fun. You can track progress [here](https://github.com/nodeca/multimath).

To summarize things: IMHO existing demo is enougth to play with. Writing more benchmarks will be not interesting waste of time for me. I prefer to do `multimath` lib, that will be more useful for all.

Of cause, if someone deside to suggest PR, that will be considered, but i think i doesn't worth to keep this issue open.

**Comment by YoranBrondsema** [comment]
Thanks for your quick reply. The reasoning makes sense.

We use pica in the context of a web application so I am interested in the differences between the various implementations on the different browsers and devices that we support. So I will still implement the benchmarks regardless :-). I'll create a PR out of it and then it'll be up to you whether you'll want to merge it, or maybe you'll just want to add the results to the README or something.

**Comment by puzrin** [comment]
To be honest, i'd prefer to avoid publishing "official" benchmarks, because such things can cause unnecessary buzz. While pica has no alternatives, having or missing benchmarks will not change too much :). But i see no problems to refer your repo with code and results.

If you are interested in technical details, i can say directly about all thin places:

- webworkers start (and probably the cause of not fast FF on mobiles)
- slow `.getPixelData()` / `.putPixelData()` (you can change nothing here)
- decode JPEG time (may be, not sure) + encode JPEG time (if you use it)

Math can be done faster for big scaledowns (~ quick 2^N scaledown + precize Lanczos3 for reminder), but:

- i did not investigated possible side effects
- IMHO currents results are good enougth :)

Also, 4x speedup possible via SIMD but we don't have it in browser now :(. And as you can see, i did all possible to be prepared for it (i think it will appear in WebAssembly first, and may be will never appear in JS).

You could try to add some euristics to WW (via pica options in constructor), but i have no ideas how to do it well without side effects. The only case i know - increase cache time if you constantly resize images on page scroll or navigation. Anyway, if you invent any good ideas after your experiments - let me know.

**Comment by puzrin** [comment]
BTW, i've fixed `cib` resize support in 3.0.3 (should enable experimental canvas features in chrome flags). It's very fast, but quality is broken. Humor is, that i've based filters logic on Skia, used un Chrome :). 

Seems bug report needed. Disabled `cib` resize by default, use it only for non-blocking image decode.

**Comment by YoranBrondsema** [comment]
OK thanks for letting me know. Damn, so they shipped a broken version of `cib` in Chrome?

I planned on working on the benchmarks this weekend.

**Comment by puzrin** [comment]
Basic support (without [options](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/createImageBitmap)) is ok. Options are described in W3C spec but not implemented anywhere except Chrome with experimental flag enabled. But as far as i see, there are problems with quality. `high` is more close to `medium`/`low`. I've added comment to [readme](https://github.com/nodeca/pica#new-picaconfig). IMHO `cib` resize now can be used only if you are ok with  "something a bit less shitty than drawImage()".

**Comment by puzrin** [comment]
https://bugs.chromium.org/p/chromium/issues/detail?id=712844 reported

**Comment by YoranBrondsema** [comment]
@puzrin Alright I have some first results. I didn't dig very deep but it was just to get an idea:

![pica](https://cloud.githubusercontent.com/assets/1374412/25465102/3cd14ff2-2ab5-11e7-9059-175118b593d3.png)

I tested on Chromium 58. As you can see, I benchmarked four sizes:

* large (3200x2500)
* medium (800x800)
* small (50x50)
* degenerate small (1x1)

Two observations:

1. The WASM version is actually slower than the JS version in most cases!
2. I added the 1x1 size to test the impact of the instantiation time of Web Workers. According to https://hacks.mozilla.org/2015/07/how-fast-are-web-workers/, the instantiation time is on the order of 40ms. This means maximum 25 operations per second. The article is from 2015 but I can't imagine it having changed orders of magnitude. However, for 1x1 we see a throughput of 567k per second, which is obviously not possible.

So before I go on, would you maybe want to go over the benchmark code and make sure it's correct? Especially 2. makes me think that it's not entirely correct. The diff is here: https://github.com/nodeca/pica/compare/master...sutori:yb-benchmarks. Thanks!

**Comment by puzrin** [comment]
1. https://github.com/nodeca/pica/compare/master...sutori:yb-benchmarks#diff-9d8e385d53cee390b1dd395abc43aa3cR25 should be `[ 'js', 'ww' ]`. Keep in mind - pica doesn't check well if options are correct or not.
2. `.resizeBuffer()` does not use all those features (see src). It's specially noted in readme, this method is for specific needs and not recommended to use if you are not 100% sure what you do.

If you need correct package benchmark, you should use `.resize()` method. Also note, `image` and `canvas` src will show different results, because jpeg src should be unpacked first. I'd recommend to check both.
